# Part 3: Translation with Flan-T5

In this part you will experiment with Google Flan-T5 models for machine translation. The original T5 model was a unifed sequence-to-sequence encoder-decoder architecture pretrained on a variety of tasks including machine translation. The Flan line of models improved on the performance of the original T5 series.

In this part we will only apply the models, not train them. We will evaluate our results using the [Bleu score](https://en.wikipedia.org/wiki/BLEU), a common (but not perfect) quantitative metric for evaluating the quality of translations. Flan-T5 also comes in several different model sizes: We will study the impact of model size on performance by considering several different versions of Flan-T5.

**Learning objectives.** You will:
1. Examine an encoder-decoder sequence-to-sequence Flan-T5 transformer model
2. Apply Flan-T5 models to perform machine translation
3. Evaluate the quality of machine translations by computing Bleu scores with respect to reference translations
4. Study the affect

While it is possible to complete this assignment using CPU compute, it may be slow. To accelerate your training, consider using GPU resources such as `CUDA` through the CS department cluster. Alternatives include Google colab or local GPU resources for those running on machines with GPU support.

First, ensure that you have the `transformers` and `datasets` modules installed. We will use these modules for importing tokenizers, pretrained models, and datasets. You can run the following cells to try to install them with `pip` if needed. If you are using ondemand, ideally you would simply include `module load transformers` and `module load datasets` when making your initial reservation.

In [1]:
pip install transformers

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install datasets

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


First we import the `flan-t5` tokenizer (shared across all model sizes) and demonstrate its characteristics and usage. Note that the API is the same as we saw previously for the `BERT` model -- may want to review the extra details in that earlier part.

Note that the example contains both English and French text.

(If you have trouble downloading the tokenizer, it is possible that you need to install the `sentencepiece` module, for example by `pip install sentencepiece`).

In [4]:
pip install sentencepiece

Defaulting to user installation because normal site-packages is not writeable
  Obtaining dependency information for sentencepiece from https://files.pythonhosted.org/packages/fb/12/2f5c8d4764b00033cf1c935b702d3bb878d10be9f0b87f0253495832d85f/sentencepiece-0.2.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 11.1 MB/s eta 0:00:0000:010:01
Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install ipywidgets --upgrade

Defaulting to user installation because normal site-packages is not writeable
  Obtaining dependency information for ipywidgets from https://files.pythonhosted.org/packages/22/2d/9c0b76f2f9cc0ebede1b9371b6f317243028ed60b90705863d493bae622e/ipywidgets-8.1.5-py3-none-any.whl.metadata
  Obtaining dependency information for comm>=0.1.3 from https://files.pythonhosted.org/packages/e6/75/49e5bfe642f71f272236b5b2d2691cf915a7283cc0ceda56357b61daa538/comm-0.2.2-py3-none-any.whl.metadata
  Obtaining dependency information for widgetsnbextension~=4.0.12 from https://files.pythonhosted.org/packages/21/02/88b65cc394961a60c43c70517066b6b679738caf78506a5da7b88ffcb643/widgetsnbextension-4.0.13-py3-none-any.whl.metadata
  Obtaining dependency information for jupyterlab-widgets~=3.0.12 from https://files.pythonhosted.org/packages/a9/93/858e87edc634d628e5d752ba944c2833133a28fa87bb093e6832ced36a3e/jupyterlab_widgets-3.0.13-py3-none-any.whl.metadata
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 139.8/139.8 k

In [3]:
# run but you do not need to modify this code

from transformers import T5Tokenizer

tokenizer = T5Tokenizer.from_pretrained("google/flan-t5-base")

print("Vocabulary size: ", tokenizer.vocab_size)
tokenized = tokenizer(["The little black cat sleeps in the window", 
                       "Le petit chat noir dort dans la fenêtre"], padding='longest')
print(tokenized)

You are using the default legacy behaviour of the <class 'transformers.models.t5.tokenization_t5.T5Tokenizer'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565


Vocabulary size:  32000
{'input_ids': [[37, 385, 1001, 1712, 2085, 7, 16, 8, 2034, 1], [312, 4561, 3582, 9691, 5048, 247, 50, 25301, 1, 0]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1, 1, 1, 0]]}


## Task 1

Below we import and preview the `flan-t5` model, beginning with the small version. You will also note that we are using 16-bit float representations to save on memory (this may be particularly relevant if you are using GPU compute for larger models, where the GPU may have limited memory available).

In [3]:
# run, but you do not need to modify this code
import torch
from transformers import T5ForConditionalGeneration

model = T5ForConditionalGeneration.from_pretrained("google/flan-t5-small", torch_dtype=torch.float32)
print(model)

2024-11-15 14:56:00.860131: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: SSE4.1 SSE4.2 AVX AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


T5ForConditionalGeneration(
  (shared): Embedding(32128, 512)
  (encoder): T5Stack(
    (embed_tokens): Embedding(32128, 512)
    (block): ModuleList(
      (0): T5Block(
        (layer): ModuleList(
          (0): T5LayerSelfAttention(
            (SelfAttention): T5Attention(
              (q): Linear(in_features=512, out_features=384, bias=False)
              (k): Linear(in_features=512, out_features=384, bias=False)
              (v): Linear(in_features=512, out_features=384, bias=False)
              (o): Linear(in_features=384, out_features=512, bias=False)
              (relative_attention_bias): Embedding(32, 6)
            )
            (layer_norm): T5LayerNorm()
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (1): T5LayerFF(
            (DenseReluDense): T5DenseGatedActDense(
              (wi_0): Linear(in_features=512, out_features=1024, bias=False)
              (wi_1): Linear(in_features=512, out_features=1024, bias=False)
              (wo): 

Examine the `model.parameters()`. How much memory (in kilobytes (KB), megabytes (MB), or gigabytes (GB)) should it take to store the model itself, given the 16-bit (or 2 byte) precision specified in the import? Briefly explain.

In [4]:
total_params = sum(p.numel() for p in model.parameters())
print(total_params)

76961152


In the case of the 2 byte precision specified in the import, the memory in total bytes is the number of parameters * bytes per parameter, which is 76,961,152 * 2, approximately 154,000,000 bytes, or about 146.5 MB (dividing 154,000,000 by 1024 * 1024). 

## Task 2

Now we import and demonstrate the basic usage of the model `generate` method. This method autoregressively generates new text as we have discussed before in the context of causal language modeling, and supports different approaches (greedy, beam search, and sampling). The `generate` method API is [documented here](https://huggingface.co/docs/transformers/v4.46.0/en/main_classes/text_generation#transformers.GenerationMixin.generate).

The example below demonstrates encoding a *batch* of inputs and passing them to the model for autoregressive generation. The printed output is generated by the model for the first and second input in the batch respectively.

In [5]:
# run but you do not need to modify this code

input_text = ["The little black cat sleeps in the window", "The dog runs in the field"]
encoded = tokenizer(input_text, return_tensors="pt", padding="longest")

outputs = model.generate(**encoded, max_new_tokens=100)
print(outputs)

for out in outputs:
    print(tokenizer.decode(out, skip_special_tokens=True))

tensor([[   0,   37, 1712,   19,    3,    9, 1712,    5,    1,    0],
        [   0,   37, 1782,   19, 1180,   16,    8, 1057,    5,    1]])
The cat is a cat.
The dog is running in the field.


Note that the model did not translate the inputs. That is by design: Flan-T5 was pretrained on several different tasks including but not limited to machine translation.

In order to use the model for translation, we need to **prompt it** to do so, providing context (literally) connecting to its pretraining. There are several different promptings that should work, we demonstrate two below:

In [16]:
# run but you do not need to modify this code

input_text = ["The little black cat sleeps in the window", "The dog runs in the field"]
prompt = "Translate English into French: "
prompted_text = [prompt + in_text for in_text in input_text]
encoded = tokenizer(prompted_text, return_tensors="pt", padding="longest")

outputs = model.generate(**encoded, max_new_tokens=100)

for out in outputs:
    print(tokenizer.decode(out, skip_special_tokens=True))

La petite cat noir s'est en vertu de la fenêtre.
Le chien s'est en l'égard de la field.


For this task, your goal is use the model to translate a large collection of German text into English, drawing from a paired translation of the novel Jane Eyre. Below we download and prepare the data.

In [4]:
# run but you do not need to modify this code

from datasets import load_dataset
from torch.utils.data import Dataset

class OpusDataset(Dataset):
    def __init__(self, dataset_stream, num_examples):
        # Convert streaming dataset to list for random access
        self.examples = list(dataset_stream.take(num_examples))
        
    def __len__(self):
        return len(self.examples)
    
    def __getitem__(self, idx):
        return self.examples[idx]

# Load the dataset in streaming mode
dataset_stream = load_dataset(
    "opus_books",
    "de-en",
    split="train",
    streaming=True
)

dataset = OpusDataset(dataset_stream, num_examples=500)

for i in range(100, 103):
    print(f"\nExample {i+1}:")
    print(f"German: {dataset[i]['translation']['de']}")
    print(f"English: {dataset[i]['translation']['en']}")
print(f"\nTotal examples loaded: {len(dataset)}")


Example 101:
German: Dieser Vorwurf meiner Abhängigkeit war in meinen Ohren fast zum leeren, bedeutungslosen Singsang geworden, sehr schmerzlich und bedrückend, aber nur halb verständlich.
English: This reproach of my dependence had become a vague sing-song in my ear: very painful and crushing, but only half intelligible.

Example 102:
German: Nun fiel auch Miß Abbot ein: »Und Sie sollten auch nicht denken, daß Sie mit den Fräulein Reed und Mr. Reed auf gleicher Stufe stehen, weil Mrs. Reed Ihnen gütig erlaubt, mit ihren Kindern erzogen zu werden.
English: Miss Abbot joined in-- "And you ought not to think yourself on an equality with the Misses Reed and Master Reed, because Missis kindly allows you to be brought up with them.

Example 103:
German: Diese werden einmal ein großes Vermögen haben, und Sie sind arm. Sie müssen demütig und bescheiden sein und versuchen, sich den andern angenehm zu machen.«
English: They will have a great deal of money, and you will have none: it is your pl

**Use the Flan-T5 model to translate all of the German text in the dataset into English.** Even with GPU compute, this may take several minutes, but should not take hours. We encourage you to add some output every 10 or 50 examples so that you can track the progress, though you are not required to do so.

**Select at least three examples from the dataset and print the model translation as well as the real English text.**

In [5]:
from datasets import load_dataset
from torch.utils.data import Dataset
import torch
from transformers import T5Tokenizer, T5ForConditionalGeneration

# Load tokenizer and model
tokenizer = T5Tokenizer.from_pretrained("google/flan-t5-base")
model = T5ForConditionalGeneration.from_pretrained("google/flan-t5-base")
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = model.to(device)

class OpusDataset(Dataset):
    def __init__(self, dataset_stream, num_examples):
        self.examples = list(dataset_stream.take(num_examples))

    def __len__(self):
        return len(self.examples)

    def __getitem__(self, idx):
        return self.examples[idx]

dataset_stream = load_dataset("opus_books", "de-en", split="train", streaming=True)
dataset = OpusDataset(dataset_stream, num_examples=500)

translations = []
batch_size = 10  

for i in range(0, len(dataset), batch_size):
    batch = dataset[i:i + batch_size]
    german_texts = [f"Translate German to English: {example['translation']['de']}" for example in batch]
    
    encoded = tokenizer(german_texts, return_tensors="pt", padding="longest", truncation=True).to(device)
    
    outputs = model.generate(**encoded, max_new_tokens=100)
    batch_translations = [tokenizer.decode(output, skip_special_tokens=True) for output in outputs]
    
    translations.extend(batch_translations)
    
    if (i + batch_size) % 50 == 0 or (i + batch_size) >= len(dataset):
        print(f"Translated {min(i + batch_size, len(dataset))}/{len(dataset)} examples.")

for i in [100, 200, 300]:
    print(f"\nExample {i+1}:")
    print(f"German: {dataset[i]['translation']['de']}")
    print(f"Real English: {dataset[i]['translation']['en']}")
    print(f"Model Translation: {translations[i]}")


Translated 50/500 examples.
Translated 100/500 examples.
Translated 150/500 examples.
Translated 200/500 examples.
Translated 250/500 examples.
Translated 300/500 examples.
Translated 350/500 examples.
Translated 400/500 examples.
Translated 450/500 examples.
Translated 500/500 examples.

Example 101:
German: Dieser Vorwurf meiner Abhängigkeit war in meinen Ohren fast zum leeren, bedeutungslosen Singsang geworden, sehr schmerzlich und bedrückend, aber nur halb verständlich.
Real English: This reproach of my dependence had become a vague sing-song in my ear: very painful and crushing, but only half intelligible.
Model Translation: This draft of my own was quickly becoming a painful and painful singing, very painful and painful, but only half a logical one.

Example 201:
German: Welche seltene Höflichkeit!
Real English: Wonderful civility this!
Model Translation: What rare happiness!

Example 301:
German: »Ich weiß es nicht.
Real English: "I don't know.
Model Translation: "I don't know.


## Task 3

Translations are difficult to evaluate quantitatively and without expert human translators. One common metric is the [BLEU score](https://en.wikipedia.org/wiki/BLEU).

The below example demonstrates calculating BLEU scores with the `evaluate` module from Hugging Face. You can [see the documentation here](https://huggingface.co/spaces/evaluate-metric/bleu). The first value in the results dictionary gives the score. Normally the score is reported on a scale from 0-100; this implementation reports it on a 0-1 scale. Higher values are better, but scores of 1 are not necessarily expected given the many possible ways to translate.

Note that `predictions` is a list of strings, but `references` is a list of lists of strings. This is because a single predicted translation could potentially have multiple equally good reference translations. In our case however we just have the single translation per pair, so `references` will be a list of lists, each with a single element.

In [23]:
pip install evaluate

Defaulting to user installation because normal site-packages is not writeable
  Obtaining dependency information for evaluate from https://files.pythonhosted.org/packages/a2/e7/cbca9e2d2590eb9b5aa8f7ebabe1beb1498f9462d2ecede5c9fd9735faaf/evaluate-0.4.3-py3-none-any.whl.metadata
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.0/84.0 kB 2.6 MB/s eta 0:00:00
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Note: you may need to restart the kernel to use updated packages.


In [22]:
# run but you do not need to modify this code

import evaluate

predictions = ["the black cat is sleeping in the sun by the window", 
               "the dog runs in the field while it rains"]
references = [["the black cat is sleeps on the sun by the window"], 
              ["the dog run in the field while it rain"]]

bleu = evaluate.load("bleu")
results = bleu.compute(predictions=predictions, references=references)
print(results)

{'bleu': 0.5555238068023582, 'precisions': [0.8, 0.6666666666666666, 0.5, 0.35714285714285715], 'brevity_penalty': 1.0, 'length_ratio': 1.0, 'translation_length': 20, 'reference_length': 20}


**Calculate the `BLEU` score of your translations from task 2 against the real English text`. Report your results.**

In [24]:
predictions = translations
references = [[example['translation']['en']] for example in dataset]

results = bleu.compute(predictions=predictions, references=references)
print(f"BLEU Score: {results['bleu']}")


BLEU Score: 0.15768461956270785


## Task 4



In this task, study the impact of model scale on the quality of the resulting translations. Earlier we used `model = T5ForConditionalGeneration.from_pretrained("google/flan-t5-small", torch_dtype=torch.float16)` to import a `flan-t5-small` model. 

**Use two additional models: `flan-t5-base` and `flan-t5-large` to generate translations of the same dataset. Evaluate and report the BLEU score of both translations.**

In [19]:
import torch
from transformers import T5ForConditionalGeneration, T5Tokenizer
import evaluate

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

bleu = evaluate.load("bleu")

model_names = ["google/flan-t5-small", "google/flan-t5-base", "google/flan-t5-large"]

# Dictionary to store BLEU scores for each model
bleu_scores = {}

for model_name in model_names:
    print(f"Processing model: {model_name}")

    tokenizer = T5Tokenizer.from_pretrained(model_name)
    model = T5ForConditionalGeneration.from_pretrained(model_name).to(device)
    
    translations = []
    for i, example in enumerate(dataset):
        german_text = example['translation']['de']
        input_text = f"Translate German to English: {german_text}"
        
        encoded = tokenizer(input_text, return_tensors="pt", padding=True, truncation=True).to(device)
        
        output = model.generate(**encoded, max_new_tokens=100)
        translated_text = tokenizer.decode(output[0], skip_special_tokens=True)
        translations.append(translated_text)
    
    references = [[example['translation']['en']] for example in dataset]
    
    results = bleu.compute(predictions=translations, references=references)
    bleu_scores[model_name] = results['bleu']
    print(f"BLEU Score for {model_name}: {results['bleu']}")

print("\nBLEU Scores for Different Models:")
for model_name, score in bleu_scores.items():
    print(f"{model_name}: {score}")


Processing model: google/flan-t5-small
BLEU Score for google/flan-t5-small: 0.07267746503411557
Processing model: google/flan-t5-base
BLEU Score for google/flan-t5-base: 0.1226615702011814
Processing model: google/flan-t5-large
BLEU Score for google/flan-t5-large: 0.15768461956270785

BLEU Scores for Different Models:
google/flan-t5-small: 0.07267746503411557
google/flan-t5-base: 0.1226615702011814
google/flan-t5-large: 0.15768461956270785
